addons/mail/models/discuss/discuss_channel.py

```diff
@@ -1200,13 +1200,23 @@ def _channel_last_message_ids(self):
         if not self:
             return []
         self.env['mail.message'].flush_model()
-        self.env.cr.execute("""
-            SELECT res_id AS id, MAX(id) AS message_id
-            FROM mail_message
-            WHERE model = 'discuss.channel' AND res_id IN %s
-            GROUP BY res_id
-            """, (tuple(self.ids),))
-        return self.env.cr.dictfetchall()
+        self.env.cr.execute(
+            """
+                   SELECT ARRAY_AGG(discuss_channel.id),
+                          ARRAY_AGG(last_message_id)
+                     FROM discuss_channel
+        LEFT JOIN LATERAL (
+                              SELECT MAX(id)
```

Comment:
I don't understand the `MAX`, why not `ORDER BY id DESC LIMIT 1` instead?

Reply:
It is not equivalent:
* `MAX` need to compare all the rows and take the biggest one
* `ORDER BY ... LIMIT 1` just needs to look at the first value of the index

This is the whole point of rewriting the query :smile:

Reply:
The explanation is basically the same as here
https://github.com/odoo/odoo/pull/161493#discussion_r1567518840

Reply:
From my testing it made no difference, the important point is the lateral join that helps the planner in this case.

Reply:
Did you test with long running channels or just at random?
In any case, it can only be better (again, as long as the index `mail_message_model_res_id_id_idx` exists, but that's a requirement for many other things perf-wise)

Reply:
Many channels, many messages per channel, and at least one channel with even more messages.

lateral order: https://explain.dalibo.com/plan/f85dfd1h8b16c8h1#plan
lateral max: https://explain.dalibo.com/plan/c55cd13106ce5dde#plan
Before PR for reference: https://explain.dalibo.com/plan/29e460261b6g17d7

This is pretty much the same in terms of complexity. Both use the same "index only scan backwards".

Maybe 0.2ms difference in favor of order by, so I'm ok to go with it, but the main goal was to go from 350ms+ to 1ms, and both would achieve that!

Reply:
Ok, from the EXPLAIN, postgres is actually doing a LIMIT implicitly as an optimisation to MAX.
Good to know 👍